autogpt_platform/backend/backend/blocks/llm.py

```diff
@@ -238,6 +242,190 @@ class Message(BlockSchema):
     content: str
 
 
+class LLMResponse(BaseModel):
+    prompt: str
+    response: str
+    tool_calls: Optional[List[ChatCompletionMessageToolCall]] | None
+    prompt_tokens: int
+    completion_tokens: int
+
+
+def llm_call(
+    credentials: APIKeyCredentials,
+    llm_model: LlmModel,
+    prompt: list[dict],
+    json_format: bool,
+    max_tokens: int | None,
+    tools: list[dict] | None = None,
+    ollama_host: str = "localhost:11434",
+) -> LLMResponse:
+    """
+    Args:
+        credentials: The API key credentials to use.
+        llm_model: The LLM model to use.
+        prompt: The prompt to send to the LLM.
+        json_format: Whether the response should be in JSON format.
+        max_tokens: The maximum number of tokens to generate in the chat completion.
+        ollama_host: The host for ollama to use
+
+    Returns:
+        The prompt sent to the LLM.
+        The response from the LLM.
+        The number of tokens used in the prompt.
+        The number of tokens used in the completion.
+    """
+    provider = llm_model.metadata.provider
+    max_tokens = max_tokens or llm_model.max_output_tokens or 4096
+
+    if provider == "openai":
+        tools_param = tools if tools else openai.NOT_GIVEN
+        oai_client = openai.OpenAI(api_key=credentials.api_key.get_secret_value())
+        response_format = None
+
+        if llm_model in [LlmModel.O1_MINI, LlmModel.O1_PREVIEW]:
+            sys_messages = [p["content"] for p in prompt if p["role"] == "system"]
+            usr_messages = [p["content"] for p in prompt if p["role"] != "system"]
+            prompt = [
+                {"role": "user", "content": "\n".join(sys_messages)},
+                {"role": "user", "content": "\n".join(usr_messages)},
+            ]
+        elif json_format:
+            response_format = {"type": "json_object"}
+
+        response = oai_client.chat.completions.create(
+            model=llm_model.value,
+            messages=prompt,  # type: ignore
+            response_format=response_format,  # type: ignore
+            max_completion_tokens=max_tokens,
+            tools=tools_param,  # type: ignore
```

Comment:
slapping `# type: ignore` on everything is probably a bad idea

Reply:
message and response_format already had a type ignore